### PR TITLE
dd-2-registration-page

### PR DESCRIPTION
**Pull request by YuliyaVoronovich**
# Task

[Registration page](https://divas801801.atlassian.net/jira/software/projects/DD/boards/2?selectedIssue=DD-2)

# Description

- Add Registration form
- Add Validators for errors

## Pull request type

Please check the type of change your PR introduces:

- [x] 🍕 Feature
- [ ] 🐛 Bugfix
- [ ] 🧑‍💻 Code Refactor
- [ ] 📝 Documentation Update
- [ ] 🤖 Build related changes
- [ ] 🔥 Performance Improvements
- [ ] Other (please describe):

# Demo (situationally)

image

# Has It Been Tested?

- [x] Yes
- [ ] No, and this is why: please replace this line with details on why tests have not been included
- [ ] Tests are not needed for this task
- [ ] I need help with writing tests

### CODE REVIEW

File: src/app/features/auth/validators/password.validator.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import { AbstractControl, FormGroup, ValidationErrors, ValidatorFn } from '@angular/forms';
+
+export function passwordValidator(): ValidatorFn {
+  return (control: AbstractControl): ValidationErrors | null =>
+    !/.{8,}/.test(control.value.trim()) ? { minLength: 'At least 8 characters required' } : null;
```
Comment (TELEUZI) [inline src/app/features/auth/validators/password.validator.ts]: Why do you need regex if you can use `length` property of string?

File: src/app/core/services/alert/alert.service.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  providedIn: 'root',
+})
+export class AlertService {
+  private readonly alert = inject(TuiAlertService);
+
+  private readonly destroy = inject(DestroyRef);
+
+  open({ message, label, appearance = 'info' }: Alert): Subscription {
+    return this.alert.open(message, { label, appearance }).pipe(takeUntilDestroyed(this.destroy)).subscribe();
+  }
```
Comment (TELEUZI) [inline src/app/core/services/alert/alert.service.ts]: Explain please how `takeUntilDestroyed` would work in this case

File: src/app/features/auth/services/auth-api/auth-api.service.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  providedIn: 'root',
+})
+export class AuthApiService {
+  private readonly httpClient = inject(HttpClient);
+
+  signup(body: UserRequest): Observable<UserResponse> {
+    return this.httpClient.post<UserResponse>('/api/signup', body).pipe(switchMap(() => this.signin(body)));
+  }
+
+  signin(body: UserRequest) {
```
Comment (TELEUZI) [inline src/app/features/auth/services/auth-api/auth-api.service.ts]: Where are modifiers? Use `typescript-eslint` rule to make it strict

File: src/app/features/auth/pages/registration-page/registration-page.component.ts [hunk 4]
```diff
@@ -1,11 +1,78 @@
...
+  registrationForm: FormGroup = this.fb.group({
+    email: this.fb.control(''),
+    password: this.fb.control(''),
+    confirmPassword: this.fb.control(''),
+    name: this.fb.control(''),
+  });
+
+  onSubmit() {
+    const emailControl = this.registrationForm.get('email');
+    const passwordControl = this.registrationForm.get('password');
```
Comment (TELEUZI) [inline src/app/features/auth/pages/registration-page/registration-page.component.ts]: Use `controls.email` etc. to not use `?` in `emailControl?.setValidators`

File: src/app/features/auth/pages/registration-page/registration-page.component.ts [hunk 5]
```diff
@@ -1,11 +1,78 @@
...
+    this.registrationForm.setValidators([matchPasswordsValidator('password', 'confirmPassword')]);
+
+    this.registrationForm.markAllAsTouched();
+
+    if (!this.registrationForm.valid) {
+      return;
+    }
+
+    const body = this.registrationForm.value;
+    this.authService.signup(body);
```
Comment (TELEUZI) [inline src/app/features/auth/pages/registration-page/registration-page.component.ts]: I suggest to return subscription from `signup` here and unsubscribe on component destroy.

File: src/app/features/auth/pages/registration-page/registration-page.component.ts [hunk 6]
```diff
@@ -1,11 +1,78 @@
...
+    const passwordControl = this.registrationForm.get('password');
+
+    this.isSubmitted = true;
+    emailControl?.setValidators([Validators.required, emailValidator()]);
+    emailControl?.updateValueAndValidity();
+    passwordControl?.setValidators([Validators.required, Validators.minLength(8)]);
+    passwordControl?.updateValueAndValidity();
+    this.registrationForm.setValidators([matchPasswordsValidator('password', 'confirmPassword')]);
+
+    this.registrationForm.markAllAsTouched();
```
Comment (TELEUZI) [inline src/app/features/auth/pages/registration-page/registration-page.component.ts]: I suggest to make `isSubmitted` signal and put this logic into effect, for example. Also, check if value is already true to not execute this logic on every submit.

File: src/app/features/auth/services/auth-api/auth-api.service.ts [hunk 7]
```diff
@@ -2,14 +2,19 @@ import { HttpClient } from '@angular/common/http';
...
 @Injectable({
   providedIn: 'root',
 })
 export class AuthApiService {
-  private readonly httpClient: HttpClient = inject(HttpClient);
+  private readonly httpClient = inject(HttpClient);
 
-  public signin(body: UserRequest) {
+  signup(body: UserRequest): Observable<UserResponse> {
+    return this.httpClient.post<UserResponse>('/api/signup', body).pipe(switchMap(() => this.signin(body)));
```
Comment (TELEUZI) [inline src/app/features/auth/services/auth-api/auth-api.service.ts]: `switchMap` is a part of business logic, it shouldn't be in `ApiService`

**Review by RykovaMariia — APPROVED** [review]

File: src/app/features/auth/pages/login-page/login-page.component.ts [hunk 8]
```diff
@@ -28,42 +28,49 @@ import { TuiValidator } from '@taiga-ui/cdk';
...
+      useValue: buildInErrors,
     },
   ],
   templateUrl: './login-page.component.html',
   styleUrl: './login-page.component.scss',
   changeDetection: ChangeDetectionStrategy.OnPush,
 })
 export class LoginPageComponent {
-  public isSubmitted = false;
+  public isSubmitted = signal(false);
```
Comment (Dewlar) [inline src/app/features/auth/pages/login-page/login-page.component.ts]: should we make the signal private and use `asReadonly` for public access ?

**Review by Dewlar — APPROVED** [review]

**Review by ChickenMcNuggetss — APPROVED** [review]

**Pull request merged**